Display file name and file size in log to easier debug which is loaded by FileInputPlugin

sakama opened the pull request:
Related with #1019

This feature is originally coming from some Embulk users.
I heard they would like to know imported file name and file size for debugging in case of file load(or parse) failure.

Some plugins like embulk-input-s3 has such logic https://github.com/embulk/embulk-input-s3/pull/68
Such logic is better to have in Embulk core.

sakama commented:
This implementation generally works for now.
```
2018-11-29 14:49:52.768 +0900 [INFO] (0015:task-0000): Loading file [/path/to/sample_01.csv], with size [238] bytes
2018-11-29 14:49:52.768 +0900 [INFO] (0016:task-0001): Loading file [/path/to/sample_02.csv], with size [280] bytes
```

However, when loading multiple files, it lacks some information.
```
[INFO] (0015:task-0000): Loading file [-], with size [238] bytes # Lack file name
[INFO] (0016:task-0001): Loading file [/path/to/sample_02.csv], with size [0] bytes
```

```
[INFO] (0015:task-0000): Loading file [/path/to/sample_01.csv], with size [238] bytes
[INFO] (0016:task-0001): Loading file [-], with size [0] bytes # lack both of file name and expected size
```

It might be related with timing of open(or close) files.
I couldn't detect the cause why such behavior happens.

dmikurube commented:
Although I'm not very sure of the cause of this, I thought we can implement this more simply.

Once we have only the following change from `master` :

```
diff --git a/embulk-core/src/main/java/org/embulk/spi/FileInput.java b/embulk-core/src/main/java/org/embulk/spi/FileInput.java
index 7a716389..d10b2b63 100644
--- a/embulk-core/src/main/java/org/embulk/spi/FileInput.java
+++ b/embulk-core/src/main/java/org/embulk/spi/FileInput.java
@@ -1,9 +1,15 @@
 package org.embulk.spi;
 
+import java.util.Optional;
+
 public interface FileInput extends AutoCloseable {
     boolean nextFile();
 
     Buffer poll();
 
     void close();
+
+    default Optional<String> hintOfInputFileNameForLogging() {
+        return Optional.empty();
+    }
 }
diff --git a/embulk-standards/src/main/java/org/embulk/standards/LocalFileInputPlugin.java b/embulk-standards/src/main/java/org/embulk/standards/LocalFileInputPlugin.java
index a768426b..dae85868 100644
--- a/embulk-standards/src/main/java/org/embulk/standards/LocalFileInputPlugin.java
+++ b/embulk-standards/src/main/java/org/embulk/standards/LocalFileInputPlugin.java
@@ -119,6 +119,11 @@ public class LocalFileInputPlugin implements FileInputPlugin {
             public TaskReport commit() {
                 return Exec.newTaskReport();
             }
+
+            @Override
+            public Optional<String> hintOfInputFileNameForLogging() {
+                return Optional.ofNullable(file.getAbsolutePath());
+            }
         };
     }
```

I guess the logging statement below works. (It's code just for testing -- not for production.)

```
diff --git a/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java b/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java
index fb2d079f..f56a545a 100644
--- a/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java
+++ b/embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java
@@ -131,8 +131,10 @@ public class FileInputRunner implements InputPlugin, ConfigurableGuessInputPlugi
         List<DecoderPlugin> decoderPlugins = newDecoderPlugins(task);
         ParserPlugin parserPlugin = newParserPlugin(task);
 
-        TransactionalFileInput tran = PluginWrappers.transactionalFileInput(
-                fileInputPlugin.open(task.getFileInputTaskSource(), taskIndex));
+        final TransactionalFileInput fromFileInputPlugin = fileInputPlugin.open(task.getFileInputTaskSource(), taskIndex);
+        Exec.getLogger(FileInputRunner.class).info("Loading file [{}]", fromFileInputPlugin.hintOfInputFileNameForLogging().orElse("-"));
+        TransactionalFileInput tran = PluginWrappers.transactionalFileInput(fromFileInputPlugin);
+
         try (CloseResource closer = new CloseResource(tran)) {
             try (AbortTransactionResource aborter = new AbortTransactionResource(tran)) {
                 FileInput fileInput = Decoders.open(decoderPlugins, task.getDecoderTaskSources(), tran);
```

Then, IIUC, we'll need to implement `hintOfInputFileNameForLogging` everywhere creating new `FileInput` from `FileInput`, for example `PluginWrappers.transactionalFileInput`, decoder plugins, and so on. What do you think?

* Although #1019 includes `expectedSize`, I think we only need filenames for the time being. Let's start just from filenames.
* To emphasize that this filename information is just a hint, not reliable "data" to be stored in columns, we may want to name it longer, redundant, and clear that it's a hint. Then I named it `hintOfInputFileNameForLogging`, but it's up to you.

kamatama41 commented:
I guess here might be the cause.
https://github.com/embulk/embulk/pull/1066/files#diff-1c8d2dffe548c7a62499cbc9ce46ceadR28
Interface's variable becomes static (and final), so the `report` would be shared with input tasks.

ref: https://www.javahelps.com/2015/02/introduction-to-interface-with-java-8.html

sakama commented:
Thanks both.

I reviced implementation in 9973a40 based on @dmikurube's comment https://github.com/embulk/embulk/pull/1066#issuecomment-442727111

I put logging logic after parserPlugin.run() in FileInputRunner class.
```
  parserPlugin.run(task.getParserTaskSource(), schema, fileInput, output);
+ Exec.getLogger(FileInputRunner.class).info(
+         "Loading file [{}] with size [{}] bytes",
+         fromFileInputPlugin.hintOfInputFileNameForLogging().orElse("-"),
+         fromFileInputPlugin.hintOfInputFileExpectedSize().orElse(new Integer(0).longValue())
+ );
```

Although LocalFileInput plugin works even if I put above logic before try block, external FileInput plugin doesn't works with this.
https://github.com/embulk/embulk/pull/1066/commits/9973a40bbf2cd20d8c9c3eb092be6f9ba78da29b#diff-f603ed6fcceb45ec6f6e43f81d52c0f6R135


In ef55c56, I added similar methods for external FileInput plugins.
External FileInput plugin like S3, GCS, SFTP get remote files in `SingleFileProvider` class that implements `InputStreamFileInput.Provider`
https://github.com/embulk/embulk-input-s3/blob/v0.2.21/embulk-input-s3/src/main/java/org/embulk/input/s3/AbstractS3FileInputPlugin.java#L443
Then, only with 9973a40, external FileInput plugins will not be able to send file name to Embulk core.

Reference implementation for embulk-input-s3 is here
https://github.com/embulk/embulk-input-s3/pull/78

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInput.java:
```diff
@@ -1,9 +1,19 @@
+        return Optional.empty();
+    }
+
+    default Optional<Long> hintOfInputFileExpectedSize() {
```
I overlooked your comment.
I removed logging logic about expected size 67af511

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java:
```diff
@@ -131,14 +131,18 @@ public TaskReport run(TaskSource taskSource, Schema schema, int taskIndex,
+                        "Loading file [{}] with size [{}] bytes",
+                        fromFileInputPlugin.hintOfInputFileNameForLogging().orElse("-"),
+                        fromFileInputPlugin.hintOfInputFileExpectedSize().orElse(new Integer(0).longValue())
+                );
```
That should be better.
There is a delicate timing issue. I can't solve this problem for now.

sakama commented on embulk-core/src/main/java/org/embulk/spi/util/InputStreamFileInput.java:
```diff
@@ -13,6 +14,24 @@
+    @Override
+    public Optional<String> hintOfInputFileNameForLogging() {
+        return provider.hintOfInputFileNameForLogging();
+    }
```
Thanks for suggestion.
This worked with LocalFileInput plugin, didn't work with external FileInput plugins like S3.

It seems logging logic is called before `provider.openNextWithHints()`. Then, file name always became null.

sakama commented:
Finally, this logic works as expected.

I tested with combination of following plugins an all of them worked fine.
* S3/Local file input plugins
* CSV/Json parser plugins
* Gzip/Bzip2 file decoder plugins

Changes are...
* Implemented `InputStreamWithHints` class for logging 9bd344c that @dmikurube mentioned at https://github.com/embulk/embulk/pull/1066#discussion_r238557396
* Moved logging logic to CSV/Json parser plugins ddf4746
  FileInput runner can't get file name since `InputStreamFileInput.nextFile()` will be called by parser plugin.
* Also implemented `InputStreamWithHints.openNextWithHints()` in each decoders 49af651
  Because parser can't get file name when we use decoder plugins.

dmikurube commented on embulk-standards/src/main/java/org/embulk/standards/CsvParserPlugin.java:
```diff
@@ -242,6 +242,8 @@ public void run(TaskSource taskSource, final Schema schema,
         try (final PageBuilder pageBuilder = new PageBuilder(Exec.getBufferAllocator(), schema, output)) {
             while (tokenizer.nextFile()) {
+                Exec.getLogger(CsvParserPlugin.class).info("Loading file [{}]", input.hintOfInputFileNameForLogging().orElse("-"));
+
```
^ And the `DataException` thrown just above the log message as well.

sakama replied:
Fixed 59cb3a0

* `stop_on_invalid_record: true`
```
2018-12-11 11:51:16.717 +0900 [WARN] (0015:task-0000): Skipped line /path/to/sample_01.csv:2 (Too many columns): 1,32864,2015-01-27 19:23:49,20150127,embulk
```

* `stop_on_invalid_record: false`
```
Error: org.embulk.spi.DataException: Invalid record at /path/to/sample_01.csv:2: 1,32864,2015-01-27 19:23:49,20150127,embulk
```

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java:
```diff
@@ -131,14 +131,13 @@ public TaskReport run(TaskSource taskSource, Schema schema, int taskIndex,
-        TransactionalFileInput tran = PluginWrappers.transactionalFileInput(
-                fileInputPlugin.open(task.getFileInputTaskSource(), taskIndex));
+        final TransactionalFileInput fromFileInputPlugin = fileInputPlugin.open(task.getFileInputTaskSource(), taskIndex);
+        TransactionalFileInput tran = PluginWrappers.transactionalFileInput(fromFileInputPlugin);
```
Fixed b10407b

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInputRunner.java:
```diff
@@ -131,14 +131,13 @@ public TaskReport run(TaskSource taskSource, Schema schema, int taskIndex,
                 FileInput fileInput = Decoders.open(decoderPlugins, task.getDecoderTaskSources(), tran);
                 closer.closeThis(fileInput);
                 parserPlugin.run(task.getParserTaskSource(), schema, fileInput, output);
-
```
Fixed b10407b

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInput.java:
```diff
@@ -1,9 +1,15 @@
 
     void close();
+
+    default Optional<String> hintOfInputFileNameForLogging() {
```
It might be long but express well about this feature.
I changed to `hintOfCurrentInputFileNameForLogging` 363bca6

sakama commented on embulk-core/src/main/java/org/embulk/spi/FileInput.java:
```diff
@@ -1,9 +1,15 @@
     Buffer poll();
 
     void close();
+
```
Added Javadoc 7d7481a

sakama commented on embulk-standards/src/main/java/org/embulk/standards/JsonParserPlugin.java:
```diff
@@ -88,6 +88,8 @@ public void run(TaskSource taskSource, Schema schema, FileInput input, PageOutpu
                 FileInputInputStream in = new FileInputInputStream(input)) {
             while (in.nextFile()) {
+                Exec.getLogger(JsonParserPlugin.class).info("Loading file [{}]", input.hintOfInputFileNameForLogging().orElse("-"));
+
```
Fixed 59cb3a0

Genrally, `DataException` will be thrown and warn log can't be used in almost cases.
```
Caused by: org.embulk.spi.DataException: Failed to parse JSON: s3://test-bucket/json/invalid_sample_01.json
```

sakama commented:
Thanks for taking a look several times. Merged.
